File: CSharp/Enterprise-Bot-Template/Middleware/Telemetry/TelemetryQnAMaker.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Collections.Generic;
using System.Net.Http;
using System.Threading.Tasks;
using Microsoft.ApplicationInsights;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.AI.QnA;

namespace EnterpriseBot
{
    /// <summary>
    /// TelemetryQnaRecognizer invokes the Qna Maker and logs some results into Application Insights.
    /// Logs the score, and (optionally) question
    /// Along with Conversation and ActivityID.
    /// The Custom Event name this logs is "QnaMessage"
    /// See <seealso cref="QnaMaker"/> for additional information.
    /// </summary>
    public class TelemetryQnAMaker : QnAMaker
    {
        public static readonly string QnaMsgEvent = "QnaMessage";

        /// <summary>
        /// Initializes a new instance of the <see cref="TelemetryQnAMaker"/> class.
        /// </summary>
        /// <param name="endpoint">The endpoint of the knowledge base to query.</param>
        /// <param name="options">The options for the QnA Maker knowledge base.</param>
        /// <param name="logUserName">The flag to include username in logs.</param>
        /// <param name="logOriginalMessage">The flag to include original message in logs.</param>
        /// <param name="httpClient">An alternate client with which to talk to QnAMaker.
        /// If null, a default client is used for this instance.</param>
        public TelemetryQnAMaker(QnAMakerEndpoint endpoint, QnAMakerOptions options = null, bool logUserName = false, bool logOriginalMessage = false, HttpClient httpClient = null)
            : base(endpoint, options, httpClient)
        {
            LogUserName = logUserName;
            LogOriginalMessage = logOriginalMessage;
        }

        public bool LogUserName { get; }

        public bool LogOriginalMessage { get; }

        public new async Task<QueryResult[]> GetAnswersAsync(ITurnContext context)
        {
            // Call Qna Maker
            var queryResults = await base.GetAnswersAsync(context);

            // Find the Application Insights Telemetry Client
            if (queryResults != null && context.TurnState.TryGetValue(TelemetryLoggerMiddleware.AppInsightsServiceKey, out var telemetryClient))
            {
                var telemetryProperties = new Dictionary<string, string>();
                var telemetryMetrics = new Dictionary<string, double>();

                // Make it so we can correlate our reports with Activity or Conversation
                telemetryProperties.Add(QnATelemetryConstants.ActivityIdProperty, context.Activity.Id);
                var conversationId = context.Activity.Conversation.Id;
                if (!string.IsNullOrEmpty(conversationId))
                {
                    telemetryProperties.Add(QnATelemetryConstants.ConversationIdProperty, conversationId);
                }

                // For some customers, logging original text name within Application Insights might be an issue
                var text = context.Activity.Text;
                if (LogOriginalMessage && !string.IsNullOrWhiteSpace(text))
                {
                    telemetryProperties.Add(QnATelemetryConstants.OriginalQuestionProperty, text);
                }

                // For some customers, logging user name within Application Insights might be an issue
                var userName = context.Activity.From.Name;
                if (LogUserName && !string.IsNullOrWhiteSpace(userName))
                {
                    telemetryProperties.Add(QnATelemetryConstants.UsernameProperty, userName);
                }

                // Fill in Qna Results (found or not)
                if (queryResults.Length > 0)
                {
                    var queryResult = queryResults[0];
                    telemetryProperties.Add(QnATelemetryConstants.QuestionProperty, string.Join(",", queryResult.Questions));
                    telemetryProperties.Add(QnATelemetryConstants.AnswerProperty, queryResult.Answer);
                    telemetryMetrics.Add(QnATelemetryConstants.ScoreProperty, queryResult.Score);
                }
                else
                {
                    telemetryProperties.Add(QnATelemetryConstants.QuestionProperty, "No Qna Question matched");
                    telemetryProperties.Add(QnATelemetryConstants.AnswerProperty, "No Qna Question matched");
                }

                // Track the event
                ((TelemetryClient)telemetryClient).TrackEvent(QnaMsgEvent, telemetryProperties, telemetryMetrics);
            }

            return queryResults;
        }
    }
}

File: samples/javascript_typescript/52.enterprise-bot/src/middleware/telemetry/telemetryQnAMaker.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License

import { TelemetryClient } from "applicationinsights";
import { TurnContext } from "botbuilder";
import { QnAMaker, QnAMakerEndpoint, QnAMakerOptions, QnAMakerResult } from "botbuilder-ai";
import { QnATelemetryConstants } from "./qnaTelemetryConstants";
import { TelemetryLoggerMiddleware } from "./telemetryLoggerMiddleware";

/**
 * TelemetryQnaRecognizer invokes the Qna Maker and logs some results into Application Insights.
 * Logs the score, and (optionally) questionAlong with Conversation and ActivityID.
 * The Custom Event name this logs is "QnaMessage"
 */
export class TelemetryQnAMaker extends QnAMaker {
    public static readonly QnAMessageEvent: string = "QnaMessage";

    private readonly _logOriginalMessage: boolean;
    private readonly _logUserName: boolean;
    private _options: { top: number, scoreThreshold: number } = { top: 1, scoreThreshold: 0.3 };

    /**
     * Initializes a new instance of the TelemetryQnAMaker class.
     * @param {QnAMakerEndpoint} endpoint The endpoint of the knowledge base to query.
     * @param {QnAMakerOptions} options The options for the QnA Maker knowledge base.
     * @param {boolean} logUserName The flag to include username in logs.
     * @param {boolean} logOriginalMessage The flag to include original message in logs.
     */
    constructor(endpoint: QnAMakerEndpoint, options?: QnAMakerOptions, logUserName: boolean = false, logOriginalMessage: boolean = false) {
        super(endpoint, options);
        this._logOriginalMessage = logOriginalMessage;
        this._logUserName = logUserName;
        Object.assign(this._options, options);
    }

    /**
     * Gets a value indicating whether determines whether to log the Activity message text that came from the user.
     */
    public get logOriginalMessage(): boolean { return this._logOriginalMessage; }

    /**
     * Gets a value indicating whether determines whether to log the User name.
     */
    public get logUserName(): boolean { return this._logUserName; }

    public async getAnswersAsync(context: TurnContext): Promise<QnAMakerResult[]> {
        // Call Qna Maker
        const queryResults: QnAMakerResult[] = await super.generateAnswer(context.activity.text, this._options.top, this._options.scoreThreshold);

        // Find the Application Insights Telemetry Client
        if (queryResults && context.turnState.has(TelemetryLoggerMiddleware.AppInsightsServiceKey)) {
            const telemetryClient: TelemetryClient = context.turnState.get(TelemetryLoggerMiddleware.AppInsightsServiceKey);

            const properties: { [key: string]: string } = {};
            const metrics: { [key: string]: number } = {};

            // Make it so we can correlate our reports with Activity or Conversation
            properties[QnATelemetryConstants.ActivityIdProperty] = context.activity.id || "";
            const conversationId: string = context.activity.conversation.id;
            if (conversationId) {
                properties[QnATelemetryConstants.ConversationIdProperty] = conversationId;
            }

            // For some customers, logging original text name within Application Insights might be an issue
            const text: string = context.activity.text;
            if (this.logOriginalMessage && text) {
                properties[QnATelemetryConstants.OriginalQuestionProperty] = text;
            }

            // For some customers, logging user name within Application Insights might be an issue
            const name: string = context.activity.from.name;
            if (this.logUserName && name) {
                properties[QnATelemetryConstants.UsernameProperty] = name;
            }

            // Fill in Qna Results (found or not)
            if (queryResults.length > 0) {
                const queryResult: QnAMakerResult = queryResults[0];
                properties[QnATelemetryConstants.QuestionProperty] = Array.of(queryResult.questions).join(",");
                properties[QnATelemetryConstants.AnswerProperty] = queryResult.answer;
                metrics[QnATelemetryConstants.ScoreProperty] = queryResult.score;
            } else {
                properties[QnATelemetryConstants.QuestionProperty] = "No Qna Question matched";
                properties[QnATelemetryConstants.AnswerProperty] = "No Qna Question matched";
            }

            // Track the event
            telemetryClient.trackEvent({
                measurements: metrics,
                name: TelemetryQnAMaker.QnAMessageEvent,
                properties,
            });
        }

        return queryResults;
    }
}
